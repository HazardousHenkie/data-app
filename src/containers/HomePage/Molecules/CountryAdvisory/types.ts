interface CountryAdvisoryInterface {
    advisory: {
        score: number
        message: string
        updated: string
        source: string
    }
}

export default CountryAdvisoryInterface
