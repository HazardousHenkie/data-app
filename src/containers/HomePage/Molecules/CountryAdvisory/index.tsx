import React from 'react'

import { createSelector } from 'reselect'
import { useSelector } from 'react-redux'

import Avatar from '@material-ui/core/Avatar'
import WarningIcon from '@material-ui/icons/Warning'
import CardActions from '@material-ui/core/CardActions'
import CardHeader from '@material-ui/core/CardHeader'
import CardContent from '@material-ui/core/CardContent'
import Typography from '@material-ui/core/Typography'

import InfoMessage from 'components/Atoms/InfoMessage'
import InlineLoader from 'components/Atoms/InlineLoader'
import makeSelectCountry from 'containers/HomePage/Molecules/CountryListItem/selectors'

import { useTranslation } from 'react-i18next'

import CardStyled, {
    CardBottomTypography,
    StyledLink
} from './styledComponents'

import useCountryAdvisory from './useCountryAdvisoryHook'

const stateSelector = createSelector(makeSelectCountry(), country => ({
    country
}))

const DrawerCountryContent: React.FC = () => {
    const { t, i18n } = useTranslation('homePage')
    const { country } = useSelector(stateSelector)
    const { loading, fetchingError, countryAdvisory } = useCountryAdvisory()

    return (
        <CardStyled variant="outlined">
            {loading && <InlineLoader />}

            <CardHeader
                avatar={
                    <Avatar
                        aria-label={t(
                            'homePage:countryAdvisor.avatar',
                            'country'
                        )}
                    >
                        <WarningIcon />
                    </Avatar>
                }
                title={`${t(
                    'homePage:countryAdvisor.advisoryTitle',
                    'Advisory for'
                )}
                ${
                    i18n.language === 'en'
                        ? country.name
                        : country.translations[i18n.language]
                }`}
            />
            <CardContent>
                {fetchingError && (
                    <InfoMessage severity="error" message={fetchingError} />
                )}

                <Typography variant="body1" component="p">
                    {countryAdvisory &&
                        countryAdvisory.advisory.score &&
                        !loading && (
                            <>
                                <strong>
                                    {t(
                                        'homePage:countryAdvisor.score',
                                        'score'
                                    )}
                                    :
                                </strong>
                                {` ${countryAdvisory.advisory.score}`}
                            </>
                        )}

                    {countryAdvisory &&
                        countryAdvisory.advisory.score &&
                        !loading && <br />}

                    {countryAdvisory && !loading && (
                        <>
                            <strong>
                                {t(
                                    'homePage:countryAdvisor.message',
                                    'message'
                                )}
                                :
                            </strong>
                            {` ${countryAdvisory.advisory.message}`}
                        </>
                    )}

                    {countryAdvisory &&
                        countryAdvisory.advisory.message === '' &&
                        !loading &&
                        t(
                            'homePage:countryAdvisor.notFound',
                            'Advisory message not available.'
                        )}

                    {!countryAdvisory &&
                        !loading &&
                        !fetchingError &&
                        t(
                            'homePage:countryAdvisor.notFound',
                            'Advisory not found.'
                        )}
                </Typography>
            </CardContent>
            {countryAdvisory && !loading && (
                <CardActions>
                    <CardBottomTypography variant="body2" color="inherit">
                        {countryAdvisory.advisory.updated}

                        <StyledLink
                            href={countryAdvisory.advisory.source}
                            color="inherit"
                            target="_blank"
                        >
                            {t('homePage:countryAdvisor.source', 'source')}
                        </StyledLink>
                    </CardBottomTypography>
                </CardActions>
            )}
        </CardStyled>
    )
}

export default DrawerCountryContent
