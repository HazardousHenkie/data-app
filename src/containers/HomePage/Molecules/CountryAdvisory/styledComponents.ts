import styled from 'styled-components'

import Card from '@material-ui/core/Card'

const CardStyled = styled(Card)`
    position: relative;
`

export default CardStyled
