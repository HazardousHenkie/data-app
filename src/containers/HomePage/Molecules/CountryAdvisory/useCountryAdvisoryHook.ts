import { useState, useEffect } from 'react'

import { createSelector } from 'reselect'
import { useSelector } from 'react-redux'

import { Reducer } from 'redux'
import { useInjectReducer } from 'utils/redux-injectors'
import makeSelectCountry from 'containers/HomePage/Molecules/CountryListItem/selectors'

import request from 'utils/request'

import reducer from 'containers/HomePage/Molecules/CountryListItem/reducer'
import CountryAdvisoryInterface from './types'

const stateSelector = createSelector(makeSelectCountry(), country => ({
    country
}))

const useCountryAdvisory = () => {
    const { country } = useSelector(stateSelector)
    const [loading, setLoading] = useState<boolean>(false)
    const [fetchingError, setFetchingError] = useState<string>()
    const [countryAdvisory, setCountryAdvisory] = useState<
        CountryAdvisoryInterface
    >()

    useInjectReducer({ key: 'country', reducer: reducer as Reducer })

    useEffect(() => {
        const fetchData = async () => {
            setLoading(true)
            try {
                const result = await request(
                    `https://www.travel-advisory.info/api?countrycode=${country.alpha2Code}`
                )

                if (result) {
                    setCountryAdvisory(result.data[country.alpha2Code])
                }
            } catch (error) {
                if (error.response.status !== 404) {
                    setFetchingError(error.responseText)
                }
            }

            setLoading(false)
        }

        fetchData()
    }, [country])

    return { loading, fetchingError, countryAdvisory }
}

export default useCountryAdvisory
