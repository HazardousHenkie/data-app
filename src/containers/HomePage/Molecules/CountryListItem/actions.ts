import { action } from 'typesafe-actions'

import ActionTypes from './constants'

import CountryState from './types'

const setSelectedCountry = (country: CountryState['country']) =>
    action(ActionTypes.SET_SELECTED_COUNTRY, country)

export default setSelectedCountry
