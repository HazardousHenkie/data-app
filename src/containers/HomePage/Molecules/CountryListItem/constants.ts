import CountryState from './types'

enum ActionTypes {
    SET_SELECTED_COUNTRY = 'components/CountriesList/SET_SELECTED_COUNTRY'
}

export const CountryItem: CountryState = {
    country: {
        alpha2Code: '',
        name: '',
        nativeName: '',
        capital: '',
        region: '',
        subregion: '',
        flag: '',
        currency: '',
        population: 0,
        latlng: [0, 0],
        currencies: [{ string: '' }],
        languages: [{ string: '' }],
        translations: { string: '' }
    }
}

export default ActionTypes
