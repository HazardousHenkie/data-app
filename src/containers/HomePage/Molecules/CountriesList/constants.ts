import { CountryItem } from 'containers/HomePage/Molecules/CountryListItem/constants'
import CountriesListState from './types'

enum ActionTypes {
    GET_COUNTRIES_DATA = 'components/CountriesList/GET_COUNTRIES_DATA',
    GET_COUNTRIES_DATA_SUCCESS = 'components/CountriesList/GET_COUNTRIES_DATA_SUCCESS',
    GET_COUNTRIES_DATA_ERROR = 'components/CountriesList/GET_COUNTRIES_DATA_ERROR'
}

export const initialCountriesHeaderState: CountriesListState = {
    error: false,
    loading: false,
    data: [CountryItem.country]
}

export default ActionTypes
