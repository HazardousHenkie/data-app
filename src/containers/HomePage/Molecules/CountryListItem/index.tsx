import React from 'react'

import { Reducer } from 'redux'

import { useInjectReducer } from 'utils/redux-injectors'

import ListItem from '@material-ui/core/ListItem'
import ListItemText from '@material-ui/core/ListItemText'
import ListItemIcon from '@material-ui/core/ListItemIcon'

import { useTranslation } from 'react-i18next'

import { useDispatch, useSelector } from 'react-redux'
import { createSelector } from 'reselect'
import { makeSelectData } from 'containers/HomePage/Molecules/CountriesList/selectors'
import { CountryInterface } from './types'
import setSelectedCountry from './actions'

import reducer from './reducer'

interface CountriesListItemProps {
    setOpen: React.Dispatch<React.SetStateAction<boolean>>
    listCountry: CountryInterface
}

const key = 'country'

const stateSelector = createSelector(makeSelectData(), countries => ({
    countries
}))

const CountriesListItem: React.FC<CountriesListItemProps> = ({
    setOpen,
    listCountry
}) => {
    const dispatch = useDispatch()
    const { i18n } = useTranslation()
    const { countries } = useSelector(stateSelector)

    useInjectReducer({ key, reducer: reducer as Reducer })

    const onClickCountry = (selectedName: string) => {
        const selectedCountry = countries.find(
            country => country.name === selectedName
        )

        if (selectedCountry) {
            setOpen(false)
            dispatch(setSelectedCountry(selectedCountry))
        }
    }

    return (
        <ListItem
            data-testid="CountriesListItem"
            button
            onClick={() => onClickCountry(listCountry.name)}
        >
            <ListItemIcon data-testid="CountriesListItemIcon">
                <img
                    width="30"
                    height="20"
                    src={listCountry.flag}
                    alt={listCountry.name}
                />
            </ListItemIcon>
            <ListItemText
                data-testid="CountriesListItemText"
                primary={
                    i18n.language === 'en'
                        ? listCountry.name
                        : listCountry.translations[i18n.language]
                }
            />
        </ListItem>
    )
}

export default CountriesListItem
