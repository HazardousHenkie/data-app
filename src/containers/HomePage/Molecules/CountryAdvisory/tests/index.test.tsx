import React from 'react'
import { render } from 'utils/test-utils'

import configureStore from 'store/configureStore'
import history from 'utils/history'

import { CountryItem } from 'containers/HomePage/Molecules/CountryListItem/constants'
import setSelectedCountry from 'containers/HomePage/Molecules/CountryListItem/actions'
import useCountryAdvisoryHook from '../useCountryAdvisoryHook'
import CountryAdvisory from '../index'

jest.mock('../useCountryAdvisoryHook')

describe('<CountryAdvisory />', () => {
    let store = configureStore({}, history)

    beforeEach(() => {
        // @ts-ignore
        useCountryAdvisoryHook.mockReturnValue({
            loading: false
        })
    })

    afterEach(() => {
        store = configureStore({}, history)
    })

    it('should render like snapshot', () => {
        jest.unmock('../useCountryAdvisoryHook')
        const component = render(<CountryAdvisory />)

        expect(component).toMatchSnapshot()
    })

    it('should render CountryAdvisory', () => {
        const { getByTestId } = render(<CountryAdvisory />)

        const countryAdvisory = getByTestId('CountryAdvisory')
        expect(countryAdvisory).toBeInTheDocument()
    })

    it('should render InlineLoader when loading', () => {
        // @ts-ignore
        useCountryAdvisoryHook.mockReturnValue({
            loading: true
        })

        const { getByTestId } = render(<CountryAdvisory />)

        const inlineLoader = getByTestId('inlineLoader')
        expect(inlineLoader).toBeInTheDocument()
    })

    it("shouldn't render InlineLoader when loading", () => {
        const { queryByTestId } = render(<CountryAdvisory />)

        const inlineLoader = queryByTestId('inlineLoader')
        expect(inlineLoader).toBeFalsy()
    })

    it('should render CountryAdvisoryHeader', () => {
        const { getByTestId } = render(<CountryAdvisory />)

        const CountryAdvisoryHeader = getByTestId('CountryAdvisoryHeader')
        expect(CountryAdvisoryHeader).toBeInTheDocument()
    })

    it('should render avatar and translated label inside avatar', () => {
        const { getByLabelText } = render(<CountryAdvisory />)

        const avatarLabel = getByLabelText('country')
        expect(avatarLabel).toBeInTheDocument()
    })

    it('should render a SVG WarningIcon inside CountryAdvisoryHeader avatar', () => {
        const { getByTestId } = render(<CountryAdvisory />)

        const CountryAdvisoryHeaderSVG = getByTestId(
            'CountryAdvisoryHeader'
        ).querySelector('svg')

        expect(CountryAdvisoryHeaderSVG).toBeInTheDocument()
    })

    it('should render a advisoryText inside CountryInformationCardHeader', () => {
        const { getByText } = render(<CountryAdvisory />, {
            store
        })

        store.dispatch(
            setSelectedCountry({ ...CountryItem.country, name: 'japan' })
        )

        const advisoryText = getByText('Advisory for japan')

        expect(advisoryText).toBeInTheDocument()
    })

    it('should render InlineLoader when loading', () => {
        const errorMessage = "Something wen't wrong"
        // @ts-ignore
        useCountryAdvisoryHook.mockReturnValue({
            fetchingError: errorMessage
        })

        const { getByText } = render(<CountryAdvisory />)

        const infoMessage = getByText(errorMessage)
        expect(infoMessage).toBeInTheDocument()
    })

    it('should render CountryAdvisoryCardContent', () => {
        const { getByTestId } = render(<CountryAdvisory />)

        const CountryAdvisoryCardContent = getByTestId(
            'CountryAdvisoryCardContent'
        )

        expect(CountryAdvisoryCardContent).toBeInTheDocument()
    })

    it('should render countryAdvisoryScore as a P', () => {
        const { getByTestId } = render(<CountryAdvisory />)

        const countryAdvisoryScore = getByTestId('countryAdvisoryScore')

        expect(countryAdvisoryScore.tagName).toBe('P')
    })

    it('should render score text inside countryAdvisoryScore', () => {
        // @ts-ignore
        useCountryAdvisoryHook.mockReturnValue({
            countryAdvisory: {
                advisory: {
                    score: 7
                }
            }
        })

        const { getByText } = render(<CountryAdvisory />)

        const score = getByText('score:')

        expect(score).toBeInTheDocument()
    })

    it('should render score inside countryAdvisoryScore', () => {
        // @ts-ignore
        useCountryAdvisoryHook.mockReturnValue({
            countryAdvisory: {
                advisory: {
                    score: 'test'
                }
            }
        })

        const { getByText } = render(<CountryAdvisory />)

        const score = getByText(/test/)

        expect(score).toBeInTheDocument()
    })

    it('should render a br countryAdvisoryScore', () => {
        // @ts-ignore
        useCountryAdvisoryHook.mockReturnValue({
            countryAdvisory: {
                advisory: {
                    score: 'test'
                }
            }
        })

        const { getByTestId } = render(<CountryAdvisory />)

        const countryAdvisoryScoreBr = getByTestId(
            'CountryAdvisoryCardContent'
        ).querySelector('br')

        expect(countryAdvisoryScoreBr).toBeInTheDocument()
    })

    it('should render message label inside countryAdvisoryScore', () => {
        // @ts-ignore
        useCountryAdvisoryHook.mockReturnValue({
            countryAdvisory: {
                advisory: {}
            }
        })

        const { getByText } = render(<CountryAdvisory />)

        const message = getByText('message:')

        expect(message).toBeInTheDocument()
    })

    it('should render message inside countryAdvisoryScore', () => {
        const message = 'Everything is fine!'
        // @ts-ignore
        useCountryAdvisoryHook.mockReturnValue({
            countryAdvisory: {
                advisory: {
                    message
                }
            }
        })

        const { getByText } = render(<CountryAdvisory />)

        const score = getByText(message)

        expect(score).toBeInTheDocument()
    })

    it('should render countryAdvisor notFound message if message is empty if countryAdvisoryScore', () => {
        // @ts-ignore
        useCountryAdvisoryHook.mockReturnValue({
            countryAdvisory: {
                advisory: {
                    message: ''
                }
            }
        })

        const { getByText } = render(<CountryAdvisory />)

        const score = getByText('Advisory message not available.')

        expect(score).toBeInTheDocument()
    })

    it('should render advisoryNotFound if no advisory is empty if countryAdvisoryScore', () => {
        const { getByText } = render(<CountryAdvisory />)

        const advisoryNotFound = getByText('Advisory not found.')

        expect(advisoryNotFound).toBeInTheDocument()
    })

    it('should render CountryAdvisoryActions', () => {
        // @ts-ignore
        useCountryAdvisoryHook.mockReturnValue({
            countryAdvisory: {
                advisory: {}
            }
        })
        const { queryByTestId } = render(<CountryAdvisory />)

        const CountryAdvisoryActions = queryByTestId('CountryAdvisoryActions')
        expect(CountryAdvisoryActions).toBeInTheDocument()
    })

    it('should render updatedText inside CountryAdvisoryActions as P', () => {
        const date = '10 september 2019'
        // @ts-ignore
        useCountryAdvisoryHook.mockReturnValue({
            countryAdvisory: {
                advisory: {
                    updated: date
                }
            }
        })
        const { getByText } = render(<CountryAdvisory />)

        const updatedText = getByText(date)
        expect(updatedText).toBeInTheDocument()
        expect(updatedText.tagName).toBe('P')
    })

    it('should render the source translation and should render inside CountryAdvisoryActions', () => {
        const link = 'https://www.myfakelinkisnicedot.com'
        // @ts-ignore
        useCountryAdvisoryHook.mockReturnValue({
            countryAdvisory: {
                advisory: {
                    source: link
                }
            }
        })
        const { getByText } = render(<CountryAdvisory />)

        const updatedText = getByText('source')
        expect(updatedText).toHaveAttribute('href', link)
    })
})
