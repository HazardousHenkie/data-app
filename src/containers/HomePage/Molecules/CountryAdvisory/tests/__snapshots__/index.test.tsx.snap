// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DrawerCountryContent /> should render like snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="MuiPaper-root MuiCard-root sc-AxirZ kNDIFW MuiPaper-outlined MuiPaper-rounded"
      >
        <div
          class="sc-AxjAm hsTYAM"
          data-testid="inlineLoader"
        >
          <div
            class="MuiCircularProgress-root MuiCircularProgress-indeterminate"
            role="progressbar"
            style="width: 40px; height: 40px;"
          >
            <svg
              class="MuiCircularProgress-svg"
              viewBox="22 22 44 44"
            >
              <circle
                class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate"
                cx="44"
                cy="44"
                fill="none"
                r="20.2"
                stroke-width="3.6"
              />
            </svg>
          </div>
        </div>
        <div
          class="MuiCardHeader-root"
        >
          <div
            class="MuiCardHeader-avatar"
          >
            <div
              aria-label="country"
              class="MuiAvatar-root MuiAvatar-circle MuiAvatar-colorDefault"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M1 21h22L12 2 1 21zm12-3h-2v-2h2v2zm0-4h-2v-4h2v4z"
                />
              </svg>
            </div>
          </div>
          <div
            class="MuiCardHeader-content"
          >
            <span
              class="MuiTypography-root MuiCardHeader-title MuiTypography-body2 MuiTypography-displayBlock"
            >
              Advisory for
                
            </span>
          </div>
        </div>
        <div
          class="MuiCardContent-root"
        >
          <p
            class="MuiTypography-root MuiTypography-body1"
          />
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="MuiPaper-root MuiCard-root sc-AxirZ kNDIFW MuiPaper-outlined MuiPaper-rounded"
    >
      <div
        class="sc-AxjAm hsTYAM"
        data-testid="inlineLoader"
      >
        <div
          class="MuiCircularProgress-root MuiCircularProgress-indeterminate"
          role="progressbar"
          style="width: 40px; height: 40px;"
        >
          <svg
            class="MuiCircularProgress-svg"
            viewBox="22 22 44 44"
          >
            <circle
              class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate"
              cx="44"
              cy="44"
              fill="none"
              r="20.2"
              stroke-width="3.6"
            />
          </svg>
        </div>
      </div>
      <div
        class="MuiCardHeader-root"
      >
        <div
          class="MuiCardHeader-avatar"
        >
          <div
            aria-label="country"
            class="MuiAvatar-root MuiAvatar-circle MuiAvatar-colorDefault"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M1 21h22L12 2 1 21zm12-3h-2v-2h2v2zm0-4h-2v-4h2v4z"
              />
            </svg>
          </div>
        </div>
        <div
          class="MuiCardHeader-content"
        >
          <span
            class="MuiTypography-root MuiCardHeader-title MuiTypography-body2 MuiTypography-displayBlock"
          >
            Advisory for
                
          </span>
        </div>
      </div>
      <div
        class="MuiCardContent-root"
      >
        <p
          class="MuiTypography-root MuiTypography-body1"
        />
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
